drivers/gpio/gpio_keys_zephyr.c

```diff
@@ -0,0 +1,153 @@
+	{                                                                                          \
+		.spec = GPIO_DT_SPEC_GET(DT_DRV_INST(i), gpios),                                   \
+		.debounce_interval_ms = DT_INST_PROP(i, debounce_interval_ms),                     \
+	}
+
+struct gpio_keys_data {
+	const struct device *dev;
+	gpio_callback_handler_t callback;
+	struct gpio_callback cb_data;
+	struct k_work_delayable work;
+	int8_t pin_state;
+};
+
+/**
+ * Handle debounced gpio pin state.
+ */
+static void gpio_keys_change_deferred(struct k_work *work)
+{
+	struct gpio_keys_data *data = CONTAINER_OF(work, struct gpio_keys_data, work);
+	const struct device *dev = data->dev;
+	const struct gpio_keys_config *cfg = (struct gpio_keys_config *)dev->config;
+
+	const int new_pressed = gpio_keys_get_pin_raw(dev);
+
+	LOG_DBG("gpio_change_deferred %s pin_state=%d, new_pressed=%d", dev->name, data->pin_state,
+		new_pressed);
+
+	/* If gpio changed, invoke callback */
+	if (new_pressed != data->pin_state) {
+		data->pin_state = new_pressed;
+		LOG_DBG("Calling callback %s %d", dev->name, new_pressed);
+		data->callback(dev, &data->cb_data, BIT(cfg->spec.pin));
+	}
+}
+
+static void gpio_keys_change_call_deferred(struct gpio_keys_data *data, uint32_t msec)
+{
+	int rv = k_work_reschedule(&data->work, K_MSEC(msec));
+
+	__ASSERT(rv >= 0, "Set wake mask work queue error");
+}
+
+static void gpio_keys_interrupt(const struct device *dev, struct gpio_callback *cbdata,
+				uint32_t pins)
+{
+	ARG_UNUSED(dev); /* This is a pointer to GPIO device, use dev pointer in
+			  * cbdata for pointer to gpio_debounce device node
+			  */
+	struct gpio_keys_data *data = CONTAINER_OF(cbdata, struct gpio_keys_data, cb_data);
+	const struct gpio_keys_config *cfg = data->dev->config;
+
+	gpio_keys_change_call_deferred(data, cfg->debounce_interval_ms);
+}
+
+int gpio_keys_zephyr_enable_interrupt(const struct device *dev,
+				      gpio_callback_handler_t gpio_keys_cb)
+{
+	int retval = -ENODEV;
+	const struct gpio_keys_config *cfg = dev->config;
+	struct gpio_keys_data *data = dev->data;
+	struct gpio_callback *cb;
+	gpio_flags_t flags;
+
+	cb = &data->cb_data;
+	data->callback = gpio_keys_cb;
+
+	gpio_init_callback(cb, gpio_keys_interrupt, BIT(cfg->spec.pin));
+	gpio_add_callback(cfg->spec.port, cb);
+	flags = (GPIO_INT_EDGE_BOTH | GPIO_INT_ENABLE) & ~GPIO_INT_DISABLE;
+
+	retval = gpio_pin_interrupt_configure(cfg->spec.port, cfg->spec.pin, flags);
+
+	return retval;
+}
+
+int gpio_keys_zephyr_disable_interrupt(const struct device *dev)
+{
+	int retval = -ENODEV;
+	const struct gpio_keys_config *cfg = dev->config;
+
+	retval = gpio_pin_interrupt_configure(cfg->spec.port, cfg->spec.pin, GPIO_INT_DISABLE);
+
+	return retval;
+}
+
+static int get_pin(const struct device *dev,
+		   int (*gpio_pin_get_fn)(const struct device *, gpio_pin_t))
+{
+	int pressed = 0;
+	const struct gpio_keys_config *cfg = dev->config;
+
+	pressed = gpio_pin_get_fn(cfg->spec.port, cfg->spec.pin);
```

Comment:
This seems kind of like far too much work.

Starting from `<zephyr/drivers/gpio_keys.h>`, we make the system call `gpio_keys_get_pin()`, which ends up calling `gpio_keys_zephyr_get_pin()`, which calls `get_pin()`, to therefore make another system call `gpio_pin_get()`, we then we're in the native driver's call?

Wouldn't it be more logical to simply get the `struct gpio_driver_api *gpio_api` in this function and run something like `gpio_api->port_get_raw()`?

Reply:
I understand your concern of the multiple layers here to get to the native drivers call.  My concern is that I'll need to duplicate the masking logic and asserts of the specific `get_pin*()` calls.  I generally prefer not to duplicate code when possible.
https://github.com/zephyrproject-rtos/zephyr/blob/016f0d4fbe93dfe6a5b51ac565faf17981bf29ab/include/zephyr/drivers/gpio.h#L1225

Reply:
Technically, you could simply factor it out and call the same inline function from two locations.

Having a double syscall penalty would introduce unwanted delays, and that matters in a lot of cases..